executor/index_merge_reader.go

```diff
@@ -126,32 +124,42 @@ func (e *IndexMergeReaderExecutor) Open(ctx context.Context) error {
 	workCh := make(chan *lookupTableTask, 1)
 	fetchCh := make(chan *lookupTableTask, len(kvRangess))
 
-	e.startIndexMergeProcessWorker(ctx, workCh, fetchCh, len(kvRangess))
-	for i := 0; i < len(kvRangess); i++ {
-		e.partialWorkerWg.Add(1)
+	e.startIndexMergeProcessWorker(ctx, workCh, fetchCh)
+
+	var err error
+	var partialWorkerWg sync.WaitGroup
+	for i := 0; i < len(kvRangess) && err == nil; i++ {
+		partialWorkerWg.Add(1)
 		if e.indexes[i] != nil {
-			err := e.startPartialIndexWorker(ctx, kvRangess[i], fetchCh, i)
-			if err != nil {
-				return err
-			}
+			err = e.startPartialIndexWorker(ctx, kvRangess[i], fetchCh, i, &partialWorkerWg)
```

Comment:
Goroutine will leak if we meet error here:
- `fetchCh` will not be close because of that `partialWorkerWg.Done()` may not be triggered in `startPartialIndexWorker`.
- Then, the `IndexMergeProcessWorker` may wait for `fetchCh` forever.
- And, the goroutine `waitPartialWorkersAndCloseFetchChan` will not exit, neither.

To solve this problem, we can add a local variable `exitCh chan struct{}`, and pass it into `startIndexMergeProcessWorker `, `startPartialIndexWorker `, `startPartialTableWorker`, and `startIndexMergeTableScanWorker`. If error is not nil, we `close(exitCh)`, and all the goroutine can exit successfully if they all listen on `exitCh`.

Reply:
updated